### PR TITLE
ARROW-348: [Python] Add build-type command line option to setup.py, build CMake extensions in a build type subdirectory

### PR DESCRIPTION
**Pull request by wesm**
This also resolves ARROW-230.

### CODE REVIEW

File: python/setup.py [hunk 1]
```diff
@@ -104,13 +96,14 @@ def run(self):
...
-                      'extra arguments for CMake')] +
-                    _build_ext.user_options)
+    user_options = ([('extra-cmake-args=', None, 'extra arguments for CMake'),
+                     ('build-type=', None, 'build type (debug or release)')]
+                    + _build_ext.user_options)
 
     def initialize_options(self):
         _build_ext.initialize_options(self)
         self.extra_cmake_args = os.environ.get('PYARROW_CMAKE_OPTIONS', '')
+        self.build_type = 'debug'
```
Comment (xhochy) [inline python/setup.py]: Could you do me a favor and pull here from an environment variable, i.e. `os.environ.get('PYARROW_BUILD_TYPE', 'debug')`?

Comment (wesm) [inline python/setup.py]: No problem

Comment (wesm) [inline python/setup.py]: No problem

**Review by xhochy — APPROVED** [review]
+1, works smooth for lower case build type.  Using `Release` though triggers errors in the code doing `glob` (as the directory is always called `release` (lowercase)).

**Pull request closed without merging**